smart_open/asb.py

```diff
@@ -0,0 +1,417 @@
+        except:
+            self._size = 0
+
+        self._raw_reader = _RawReader(self._blob, self._size)
+        self._current_pos = 0
+        self._current_part_size = buffer_size
+        self._current_part = smart_open.bytebuffer.ByteBuffer(buffer_size)
+        self._eof = False
+        self._line_terminator = line_terminator
+
+        #
+        # This member is part of the io.BufferedIOBase interface.
+        #
+        self.raw = None
+
+    #
+    # Override some methods from io.IOBase.
+    #
+    def close(self):
+        """Flush and close this stream."""
+        logger.debug("close: called")
+        self._blob = None
+        self._raw_reader = None
+
+    def readable(self):
+        """Return True if the stream can be read from."""
+        return True
+
+    def seekable(self):
+        """If False, seek(), tell() and truncate() will raise IOError.
+
+        We offer only seek support, and no truncate support."""
+        return True
+
+    #
+    # io.BufferedIOBase methods.
+    #
+    def detach(self):
+        """Unsupported."""
+        raise io.UnsupportedOperation
+
+    def seek(self, offset, whence=smart_open.constants.WHENCE_START):
+        """Seek to the specified position.
+
+        :param int offset: The offset in bytes.
+        :param int whence: Where the offset is from.
+
+        Returns the position after seeking."""
+        logger.debug('seeking to offset: %r whence: %r', offset, whence)
+        if whence not in smart_open.constants.WHENCE_CHOICES:
+            raise ValueError('invalid whence, expected one of %r' % smart_open.constants.WHENCE_CHOICES)
+
+        if whence == smart_open.constants.WHENCE_START:
+            new_position = offset
+        elif whence == smart_open.constants.WHENCE_CURRENT:
+            new_position = self._current_pos + offset
+        else:
+            new_position = self._size + offset
+        self._current_pos = new_position
+        self._raw_reader.seek(new_position)
+        logger.debug('current_pos: %r', self._current_pos)
+
+        self._current_part.empty()
+        self._eof = self._current_pos == self._size
+        return self._current_pos
+
+    def tell(self):
+        """Return the current position within the file."""
+        return self._current_pos
+
+    def truncate(self, size=None):
+        """Unsupported."""
+        raise io.UnsupportedOperation
+
+    def read(self, size=-1):
+        """Read up to size bytes from the object and return them."""
+        if size == 0:
+            return b''
+        elif size < 0:
+            self._current_pos = self._size
+            return self._read_from_buffer() + self._raw_reader.read()
+
+        #
+        # Return unused data first
+        #
+        if len(self._current_part) >= size:
+            return self._read_from_buffer(size)
+
+        #
+        # If the stream is finished, return what we have.
+        #
+        if self._eof:
```

Comment:
No need to use eof variable here.

```suggestion
        if self._current_pos == self._size:
```